server/src/main/java/org/elasticsearch/cluster/metadata/DesiredNodes.java

```diff
@@ -39,6 +40,71 @@
 import static java.lang.String.format;
 import static org.elasticsearch.node.Node.NODE_EXTERNAL_ID_SETTING;
 
+/**
+ * <p>
+ *  Desired nodes represents the cluster topology that the operator of the cluster is aiming for.
+ *  Therefore, it is possible that the desired nodes contain nodes that are still not part of the
+ *  cluster in contrast to {@link DiscoveryNodes} that contains only nodes that are part of the cluster.
+ * </p>
+ *
+ * <p>
+ *  This concept is useful as it provides more context about future topology changes to the system,
+ *  allowing it to make better decisions about allocation, autoscaling, auto-expand replicas, etc.
+ * </p>
+ *
+ * <p>
+ *  Additionally, settings validation is done during desired nodes updates avoiding boot-looping
+ *  when an invalid setting is provided before the node is started.
+ * </p>
+ *
+ * <p>
+ *  To modify the desired nodes it is needed to provide the entire collection of nodes that will
+ *  be part of the proposed cluster topology.
+ * </p>
+ *
+ * <p>
+ *  Desired nodes are expected to be part of a lineage defined by the provided {@code historyId}.
+ *  The {@code historyId} is provided by the orchestrator taking care of managing the cluster.
+ *  In order to identify the different proposed desired nodes within the same history, it is
+ *  also expected that the orchestrator provides a monotonically increasing {@code version}
+ *  when it communicates about future topology changes.
+ *  The cluster would reject desired nodes updated with a {@code version} less than or equal
+ *  than the current {@code version} for the same {@code historyId}.
+ * </p>
+ *
+ * <p>
+ *  The {@code historyId} is expected to remain stable during the cluster lifecycle, but it is
+ *  possible that the orchestrator loses its own state and needs to be restored to a
+ *  previous point in time with an older desired nodes {@code version}. In those cases it is
+ *  expected to use new {@code historyId} that would allow starting from a different version.
```

Comment:
Does `historyId` need to be monotonically increasing, or will any different history id work?

Reply:
`historyId` is a string, so any different history id works in this case